classes/log_fetcher/course_log_fetcher.php

```diff
@@ -0,0 +1,235 @@
+        self::write_s3_logs_to_course_logs($courseids, $courseloghandles, $s3logkeys);
+
+        echo "Fetching logs from DB \n";
+        self::write_current_database_logs_to_course_logs($courseids, $courseloghandles);
+
+        self::close_course_log_file_handles($courseloghandles);
+
+        echo "Finished pulling logs \n";
+    }
+
+    /**
+     * Pulls s3 logs, parses them and writes relevent course log entries to
+     * respective files.
+     *
+     * @param array $courseids        Course ids.
+     * @param array $courseloghandles Course log file handles.
+     * @param array $s3logkeys        S3 log keys
+     * @return void
+     */
+    private static function write_s3_logs_to_course_logs($courseids, $courseloghandles, $s3logkeys) {
+        $currentlogcount = 0;
+        $totallogs = count($s3logkeys);
+        foreach ($s3logkeys as $s3logkey) {
+            $currentlogcount++;
+
+            echo "\nDownloading $s3logkey ($currentlogcount/$totallogs) \n";
+            $loghandle = self::download_s3_log_file($s3logkey);
+            $completedlogs = 0;
+
+            echo "Parsing $s3logkey \n";
+            while (($row = fgetcsv($loghandle)) !== false) {
+                if (self::is_course_log_entry($courseids, $row)) {
+                    $entrycourseid = $row[11]; // holds contextinstanceid
+                    fputcsv($courseloghandles[$entrycourseid], $row);
+
+                    echo "Found log entry for course $entrycourseid \n";
+
+                    if ($row[1] == '\core\event\course_deleted') {
+                        $completedlogs++;
+
+                        // We've parsed all logs needed.
+                        if ($completedlogs == count($courseids)) {
+                            echo "Finishing early, {$completedlogs} course delete events found \n";
+                            fclose($loghandle);
+                            return;
+                        }
+                    }
+                }
+            }
+
+            fclose($loghandle);
+        }
+    }
+
+    /**
+     * Determines if log row is a course log entry we want.
+     * Matches courseid = contextinstanceid and contextlevel = course
+     *
+     * @param array $courseids Course ids.
+     * @param array $row       Log row.
+     * @return boolean
+     */
+    private static function is_course_log_entry($courseids, $row) {
+        if (in_array($row[11], $courseids) && $row[10] == CONTEXT_COURSE) {
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Downloads an s3 log file to the temp dir and opens a handle for it.
+     *
+     * @param string $s3logkey
+     * @return resource
+     */
+    private static function download_s3_log_file($s3logkey) {
+        $s3client = new s3_client();
+        $tempdir = make_temp_directory('s3logs_download');
+        $tempfilepath = tempnam($tempdir, 's3logs_');
+        $s3client->download_file($tempfilepath, $s3logkey);
```

Comment:
probably shouldn't open file handle unless we have successfully got the file from s3